soc: intel_adsp: Manage power gating based on core activity

tmleman opened the pull request:
This patch enhances the power management capabilities of the Intel ADSP by ensuring that power gating states are appropriately managed based on core activity. It prevents the primary core from entering power gating if secondary cores are active and re-enables power gating when all secondary cores are off, using `pm_policy_state_lock_get` and `pm_policy_state_lock_put` functions. 

The Sound Open Firmware (SOF) project currently uses a custom power management policy to achieve these effects. With this patch, the default power management policy can be utilized, allowing the option to disable the custom policy while maintaining system reliability and performance across different core states.

tmleman committed:
soc: intel_adsp: Manage power gating based on core activity

This patch enhances the power management capabilities of the Intel ADSP
by ensuring that power gating states are appropriately managed based on
core activity. It prevents the primary core from entering power gating
if secondary cores are active and re-enables power gating when all
secondary cores are off, using pm_policy_state_lock_get and
pm_policy_state_lock_put functions.

The Sound Open Firmware (SOF) project currently uses a custom power
management policy to achieve these effects. With this patch, the default
power management policy can be utilized, allowing the option to disable
the custom policy while maintaining system reliability and performance
across different core states.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
soc: intel_adsp: Update comment style and fix typos

This patch makes cosmetic changes to ace/power.c by updating comments to
Doxygen style, fixing typos, and removing an extraneous character for
improved readability and consistency.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

softwarecki commented:
@tmleman Can you fix the `biref` typo also in `cavs/power.c`?

kv2019i commented on soc/intel/intel_adsp/ace/multiprocessing.c:
```diff
@@ -147,6 +148,11 @@ void soc_start_core(int cpu_num)
+		/* The primary core cannot enter power gating if any of the secondary cores are
+		 * active.
+		 */
+		pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
```
So each secondary gets a lock reference and when the last gives up its reference, idle is allowed, right?

tmleman replied:
Exactly.

tmleman committed:
soc: intel_adsp: Fix typo in cavs/power.c comment

This patch makes cosmetic changes to cavs/power.c by updating comments to
Doxygen style, fixing typos.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
SOF CI: https://github.com/thesofproject/sof/pull/9992
